File: app/src/main/cpp/getProp_corr.cpp
#include "getProp_corr.h"

#include <iostream>

void getProp_Corr(double X[3], double pos[3], double *iono, double tow, double &tropoCorr, double &ionoCorr){

    double PVT[3] = {pos[0], pos[1], pos[2]};

    double az, el, h;
    topocent(PVT,X,az,el,h);
    
    double llh[3];
    xyz2llh(PVT, llh);

    tropoCorr = tropo_error_correction(el,llh[2],1);
    
    ionoCorr = 0.00001;
    if(iono){
        ionoCorr = iono_error_correction(llh[0]*180/M_PI, llh[1]*180/M_PI, az, el, tow,iono, nullptr);
    }

}

File: app/src/main/java/com/inari/team/computation/corrections/GetPropCorr.kt
package com.inari.team.computation.corrections

import com.inari.team.computation.converters.ecef2lla
import com.inari.team.computation.converters.toTopocent
import com.inari.team.computation.data.EcefLocation
import com.inari.team.computation.data.PvtEcef
import com.inari.team.computation.utils.Constants.KLOBUCHAR
import com.inari.team.presentation.model.PositionParameters


data class PropCorr(
    var tropoCorr: Double = 0.0,
    var ionoCorr: Double = 0.0
)

/**
 * Gets the propagation effects corrections
 */
fun getPropCorr(
    satPos: EcefLocation,
    refPos: PvtEcef,
    iono: ArrayList<Double>,
    tow: Double,
    corrections: ArrayList<Int>
): PropCorr {

    val refPosArray = doubleArrayOf(refPos.x, refPos.y, refPos.z)
    val xPosArray = doubleArrayOf(satPos.x, satPos.y, satPos.z)

    // Coordinates transformation
    val topoSatPos = toTopocent(refPosArray, xPosArray)
    val llaRefPos = ecef2lla(EcefLocation(refPos.x, refPos.y, refPos.z))

    val tropoCorr = if (corrections.contains(PositionParameters.CORR_TROPOSPHERE)) {
        // Get tropospheric correction (Saastamoinen model)
        tropoErrorCorrection(arrayListOf(topoSatPos.elevation), arrayListOf(llaRefPos.altitude))
    } else 0.0


    val ionoCorr = if (corrections.contains(PositionParameters.CORR_IONOSPHERE)) {
        // Get ionospheric correction
        if (iono.isNotEmpty()) {
            ionoErrorCorrections(llaRefPos, topoSatPos, tow, iono, KLOBUCHAR)
        } else {
            0.00001
        }
    } else 0.0

    return PropCorr(tropoCorr, ionoCorr)
}